pintracker/stateless/stateless.go

```diff
@@ -0,0 +1,489 @@
+	case ch <- op:
+	default:
+		err := errors.New("queue is full")
+		op.SetError(err)
+		op.Cancel()
+		logger.Error(err.Error())
+		return err
+	}
+	return nil
+}
+
+// SetClient makes the StatelessPinTracker ready to perform RPC requests to
+// other components.
+func (spt *Tracker) SetClient(c *rpc.Client) {
+	spt.rpcClient = c
+	spt.rpcReady <- struct{}{}
+}
+
+// Shutdown finishes the services provided by the StatelessPinTracker
+// and cancels any active context.
+func (spt *Tracker) Shutdown() error {
+	spt.shutdownMu.Lock()
+	defer spt.shutdownMu.Unlock()
+
+	if spt.shutdown {
+		logger.Debug("already shutdown")
+		return nil
+	}
+
+	logger.Info("stopping StatelessPinTracker")
+	spt.cancel()
+	close(spt.rpcReady)
+	spt.wg.Wait()
+	spt.shutdown = true
+	return nil
+}
+
+// Track tells the StatelessPinTracker to start managing a Cid,
+// possibly triggering Pin operations on the IPFS daemon.
+func (spt *Tracker) Track(c api.Pin) error {
+	logger.Debugf("tracking %s", c.Cid)
+
+	// Trigger unpin whenever something remote is tracked
+	// Note, IPFSConn checks with pin/ls before triggering
+	// pin/rm.
+	if c.IsRemotePin(spt.peerID) {
+		op := spt.optracker.TrackNewOperation(c, optracker.OperationRemote, optracker.PhaseInProgress)
+		if op == nil {
+			return nil // ongoing unpin
+		}
+		err := spt.unpin(op)
+		op.Cancel()
+		if err != nil {
+			op.SetError(err)
+		} else {
+			op.SetPhase(optracker.PhaseDone)
+		}
+		return nil
+	}
+
+	return spt.enqueue(c, optracker.OperationPin)
+}
+
+// Untrack tells the StatelessPinTracker to stop managing a Cid.
+// If the Cid is pinned locally, it will be unpinned.
+func (spt *Tracker) Untrack(c *cid.Cid) error {
+	logger.Debugf("untracking %s", c)
+	return spt.enqueue(api.PinCid(c), optracker.OperationUnpin)
+}
+
+// StatusAll returns information for all Cids pinned to the local IPFS node.
+func (spt *Tracker) StatusAll() []api.PinInfo {
+	pininfos, err := spt.localStatus(true)
+	if err != nil {
+		logger.Error(err)
+		return nil
+	}
+
+	// get all inflight operations from optracker and
+	// put them into the map, deduplicating any already 'pinned' items with
+	// their inflight operation
+	for _, infop := range spt.optracker.GetAll() {
+		pininfos[infop.Cid.String()] = infop
+	}
+
+	var pis []api.PinInfo
+	for _, pi := range pininfos {
+		pis = append(pis, pi)
+	}
+	return pis
+}
+
+// Status returns information for a Cid pinned to the local IPFS node.
+func (spt *Tracker) Status(c *cid.Cid) api.PinInfo {
+	// check if c has an inflight operation or errorred operation in optracker
+	if oppi, ok := spt.optracker.GetExists(c); ok {
+		// if it does return the status of the operation
+		return oppi
+	}
+
+	// check global state to see if cluster should even be caring about
+	// the provided cid
+	var gpin api.PinSerial
+	err := spt.rpcClient.Call(
+		"",
+		"Cluster",
+		"PinGet",
+		api.PinCid(c).ToSerial(),
+		&gpin,
+	)
+	if err != nil {
+		if _, ok := err.(*api.CidNotInGlobalStateError); ok {
```

Comment:
As a note, the fact that this works is because of the local-rpc-shortcut part of gorpc. I think the right way to fix this is on `gorpc`, declaring custom error types there are having a function like `gorpc.IsRPCError(err)` to identify them.

Mind going that way instead?

Reply:
Yeah, that sounds like a really good solution. 👍